pkg/ingester/owned_series.go

```diff
@@ -80,30 +84,76 @@ func newOwnedSeriesService(
 		getLocalSeriesLimit: getLocalSeriesLimit,
 		getTSDBUsers:        getTSDBUsers,
 		getTSDB:             getTSDB,
+		interval:            interval,
 		ownedSeriesCheckDuration: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
 			Name:    "cortex_ingester_owned_series_check_duration_seconds",
 			Help:    "How long does it take to check for owned series for all users.",
 			Buckets: prometheus.DefBuckets,
 		}),
 	}
 
-	oss.Service = services.NewTimerService(interval, nil, oss.onPeriodicCheck, nil)
+	oss.Service = services.NewBasicService(oss.starting, oss.running, nil)
 	return oss
 }
 
-// This function runs periodically. It checks if ring has changed, and updates number of owned series for any
-// user that requires it (due to ring change, compaction, shard size change, ...).
-func (oss *ownedSeriesService) onPeriodicCheck(ctx context.Context) error {
-	ringChanged, err := oss.ringStrategy.checkRingForChanges()
+// This method should run as fast as possible and avoid blocking on external conditions
+// (e.g. whether lifecycler added instance to the ring or not),
+// because Ingester writes are not allowed until ownedSeriesService switches to Running state.
+func (oss *ownedSeriesService) starting(ctx context.Context) error {
+	// Fetch and cache current state of the ring.
+	_, err := oss.ringStrategy.checkRingForChanges()
 	if err != nil {
-		level.Error(oss.logger).Log("msg", "can't check ring for updates", "err", err)
-		return nil // If we returned error, service would stop.
-	}
+		if errors.Is(err, ring.ErrEmptyRing) {
+			level.Warn(oss.logger).Log("msg", "skipped initial owned series computation, ring is empty")
+			oss.initialRingCheckSucceeded = false
+			// Service will continue in this case.
+		} else {
+			return fmt.Errorf("can't read ring: %v", err)
+		}
+	} else {
+		oss.initialRingCheckSucceeded = true
 
-	oss.updateAllTenants(ctx, ringChanged)
+		// Check all tenants, regardless of whether instance exists in the ring or not.
+		// This runs after TSDBs are open, all tenants should have "new user" reason set.
+		_ = oss.updateAllTenants(ctx, true)
+	}
 	return nil
 }
 
+// Running function of owned series service. It regularly checks if ring has changed, and updates number of owned series for any
+// user that requires it (due to ring change, compaction, shard size change, ...).
+func (oss *ownedSeriesService) running(ctx context.Context) error {
+	tickerInterval := oss.interval
+	if !oss.initialRingCheckSucceeded {
+		tickerInterval = tickerInterval / 10
```

Comment:
Should we just hard-code it to something like `100ms`? With the default `interval` this will run every `1.5s`, but I think we want to detect a success and run the initial calculation faster than that, and tightly spinning on failure shouldn't have negative effects.